CPU Usage

ggetz opened the pull request:
Changes from PRs #6065 and #6107
Opt in fix for #1865

Created opt-in `requestRenderMode` which will only render the scene when needed. Renders can be triggered as needed by calling `requestRender()`, for example if picking a feature should change it's color, or you make changes to the scene via the API.

We render each time the simulation time change exceeds `scene.maximumRenderTimeChange`, which defaults to `0.5` seconds.  It can be set to `undefined` or `Infinity` to never change based on simulation time.

Besides simulation time changes and calls to `requestRender`, always render when:
* The camera changes
* Web worker returns, successfully or with an error
* RequestScheduler fulfills a request, successfully or not
* A `FrameState.afterRender` function was called (this is used thorough the code base when changes are made after a render to keep the current state. An example would be when a new LOD of a tileset is shown or when a model is ready)
* Globe tiles are loaded
* ImageryLayers are added, removes, shown, hidden, or moved.
* The scene morphs

In addition, I refactored some of the update/render logic in `Globe`, `QuadtreePrimitive`, and `GlobeSurfaceTileProvider` so that loading new tiles will continue to prompt render requests and finish processing the load queue. When the scene has `requestRenderMode` enabled and there is a large or undefined `maximumRenderTimeChange`, the globe will still load in completely, and changing terrain providers or imagery, and loading in higher level tiles when the camera changes will automatically trigger a render.

cesium-concierge commented:
![Signed CLA is on file.](https://img.shields.io/badge/CLA-signed-brightgreen.svg)

@ggetz, thanks for the pull request! Maintainers, we have a signed CLA from @ggetz, so you can review this at any time.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

pjcozzi commented:
* Is the plan to ship this with 1.42?  OK with me, but if so, let's aim to merge this by the end of the week so it has some time in master.
* It would be worth writing a blog or short tutorial about why this exists, how to use it, and how it works. 
* I think this needs a side-by-side example showing the default auto mode and explicitly using `requestRenderMode` in a few common scenarios so it is super clear.  Perhaps this is an additional Sandcastle example?
* Is any other near-term work needed for #1865?  Or can we close it?  I am OK with not having a fully automated `requestRenderMode` at this point.  We can get community feedback if it becomes confusing.

ggetz commented:
> Is the plan to ship this with 1.42? OK with me, but if so, let's aim to merge this by the end of the week so it has some time in master.

Yes, if possible.

> It would be worth writing a blog or short tutorial about why this exists, how to use it, and how it works.

:+1:

> I think this needs a side-by-side example showing the default auto mode and explicitly using requestRenderMode in a few common scenarios so it is super clear. Perhaps this is an additional Sandcastle example?

I can modify the current sandcastle example to show some common scenarios. I don't think we need to add additional Sandcastle examples for each scenario.

> Is any other near-term work needed for #1865? Or can we close it? I am OK with not having a fully automated requestRenderMode at this point. We can get community feedback if it becomes confusing.

IMO, any remaining work in this regard will be more involved and lower level. It will also end up being less flexible for implementing in an app (for example, you may not need to request a render on every mouse movement, only if the mouse moves over a tileset to highlight a feature, it'll depend on the app).

pjcozzi commented:
All sounds good.

ggetz commented:
@kring, @bagnell, or @shunter If one of you could take a look at my approach in separating the loading and rending in `Globe`, `QuadtreePrimitive`, and `GlobeSurfaceTileProvider`, it would be much appreciated!

ggetz commented:
Updated Sandcastle example to cover some additional scenarios.

lilleyse commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -115,26 +220,33 @@
+    text : 'Load CZML with model and set the view',
+    onselect : loadCzmlScenario
+}, {
+    text : 'Scene changes with API',
```
This mode should reset the camera to be away from the globe. Otherwise when going from `Load CZML with model and set the view` to `Scene changes with API` it looks like nothing changed.

ggetz replied:
Fixed.

lilleyse commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -115,26 +220,33 @@
+    text : 'Load a 3D tileset and set the view',
+    onselect : loadTilesetScenario
+}, {
+    text : 'Mouseover picking',
```
Maybe it's just me, but the picking feel like it has a bit of a delay. Is that expected?

ggetz replied:
I see the delay. It only happens the first time an entity is added to the scene. I need to look more to see what's causing this.

ggetz replied:
I believe this is resolved, let me know if you are still seeing the delay.

pjcozzi commented:
When the Sandcastle example loads, I expect the FPS to be N/A, but how is the 17 MS reported if the FPS is N/A?  Shouldn't the frame time also be N/A since nothing is being rendered.

It doesn't seem to make a difference if I start/stop animation.  I would expect FPS/MS to be reporting when the simulation clock is changing, but not when it is paused and nothing else needs updating.

pjcozzi commented:
This doesn't seem to change when animating:

![image](https://user-images.githubusercontent.com/782098/35051399-207760a0-fb73-11e7-8789-81eece474990.png)

pjcozzi commented:
If I click "Request a new render frame" as fast as possible I would expect it to report FPS, but it stays N/A.

pjcozzi commented:
For naming and branding in Sandcastle, the blog, and the reference doc, we might want to replace "requestRender mode enabled" with "explicit rendering."

pjcozzi commented:
I think most of the confusion for https://github.com/AnalyticalGraphicsInc/cesium/pull/6115#issuecomment-358344064, https://github.com/AnalyticalGraphicsInc/cesium/pull/6115#issuecomment-358344300, and https://github.com/AnalyticalGraphicsInc/cesium/pull/6115#issuecomment-358344445 comes down to either both FPS and MS should be reported or neither.

pjcozzi commented:
> Maximum simulation time change

What does this mean?  Delta time change that triggers a new render?

Consider renaming in the UI.

None of this will be obvious to users so we need to try to make the Sandcastle wording as friendly as possible.

pjcozzi commented:
> Maximum simulation time change

Also, this slider perhaps should have a much longer duration so that users can see the effect, e.g., the lighting updates and star movement should be choppy when it is very high like an hour or few hours.

pjcozzi commented:
This Sandcastle example still doesn't show how a user will update an app from auto rendering to explicitly making requests - this will be the main crux for users.  Maybe it is too hard to convey in the Sandcastle example.  If so, please make sure the blog post clearly covers this with examples and stresses that the user likely does not need to add dozens of `requestRender` calls - their app is at a higher level of abstraction and they likely just have one or a few spots where they need to add it.

ggetz commented:
Updated terminology and added some text for clarification in the sandcastle demo. Also tweaked the FPS display according to the above comments.

pjcozzi commented:
Sandcastle updates LGTM

hpinkos commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -0,0 +1,283 @@
+        }
+    });
+
+    // If the mouse is over the billboard, change its scale and color,
```
billboard -> box

ggetz commented:
Thanks for the feedback @kring, especially for your insight into the terrain engine structure! I'll do my best to incorporate your advice and make the render/update split cleaner.

> In the CPU fix I've been using for years now, Cesium does nothing when it's idle. In this implementation, it calls update 60 times per second, which seems wasteful.

The problem with suspending all updates is that we still need to be updating the request schedulers and processing requests. When cesium is idle, for instance, we can still have LODs from 3D tilesets loading in that need to be rendered. I suppose if we are idle, we can target a lower update rate.

ggetz commented on Source/Scene/GlobeSurfaceTileProvider.js:
```diff
@@ -252,6 +253,17 @@ define([
+         * @memberof GlobeSurfaceTileProvider.prototype
+         * @type {Event}
+         */
+        imageryLayersUpdatedEvent : {
```
Regarding 1 and 2, the scene watches for this event to trigger a new render, and aggregating the events in `ImageryLayerCollection` seemed cleaner, especially because we have to handle removing the callbacks if the globe changes. Also for 3, I'll try and set this up to be raised after the update completes.

kring replied:
Oh sorry, my claim that the event isn't used was based on search the diff, and GitHub had helpfully hidden part of it.  Nevermind!  3 is probably worth a bit of thought, but may not be a big deal.

ggetz commented on Source/Scene/QuadtreePrimitive.js:
```diff
@@ -260,7 +262,75 @@ define([
+     */
+    QuadtreePrimitive.prototype.update = function(frameState) {
+        if (defined(this._tileProvider.updateImagery)) {
+            this._tileProvider.updateImagery();
```
I agree that this is a bit awkward. I think adding an `update` to `QuadtreeTileProvider` interface is probably the more generic solution. I was hesitant to update the interface because I wan't sure of the implications as `GlobeSurfaceTileProvider` is the only class in Cesium that implements it and the naming is a bit awkward with the methods that already exist.

kring replied:
Making it optional the way you've done makes sense, so that existing code using `QuadtreePrimitive` (is there any such code anymore? 🤔) won't break. Maybe just call it `update` instead of `updateImagery`?

ggetz commented:
> In Cesium's terrain engine, loading a new tile can cause the selected set of tiles to change, and thus the load queue can and will change. This is ok because the request completing also triggers render, but it feels a bit accidental. And with completed requests triggering render, why do we need to do the update stuff at 60Hz anyway? There's no need to process the load queue; it won't result in any new actions until a completed request / task produces new data and frees up a slot.

`Scene` watches the globe `tileLoadEvent` and triggers a new render. Additionally, since it's raised inside an `afterRender` function, a new frame is also triggered, so this is not really by accident. However  I will definitely take another look at the frequency of updating the load queue.

kring commented:
> The problem with suspending all updates is that we still need to be updating the request schedulers and processing requests. When cesium is idle, for instance, we can still have LODs from 3D tilesets loading in that need to be rendered. I suppose if we are idle, we can target a lower update rate.

Are you sure? If a request loading / task completing triggers a render, that render will process the results of the request/task, render the new data, and start new requests/tasks as necessary, right?

The trick, I think, is that it may take more than one render frame from the time a request/task finishes until the time it has been completely processed and shows up on the screen. I've dealt with this possibility in the past by checking load queues and only pausing rendering if they're empty. You could improve on that by having primitives report whether they have more work to do in a future frame.  i.e. the globe would say "no work to do for now, go ahead and pause rendering" even if there are tiles in the load queue but they're all waiting on a request/task. I think that's part of the future CPU work, though, not necessarily in scope for this PR.

Let me know if that makes any sense at all. :)

kring commented:
I guess I can sum this up as: I think it's better to do full renders while load is in progress than to do a subset of render (update) at 60Hz _always_. In the former approach, idle is truly idle, 0% CPU usage, which is an important feature IMO. Then future PRs just need to work to achieve that idle state sooner, for example by having primitives report more about their loading/processing state so that we can "go idle" without waiting for the load queues to drain completely.

Not sure if you guys have spent a lot of time thinking / talking about this and I'm just derailing things, but hopefully this is some food for though at least.

pjcozzi commented:
> I guess I can sum this up as: I think it's better to do full renders while load is in progress than to do a subset of render (update) at 60Hz always. In the former approach, idle is truly idle, 0% CPU usage, which is an important feature IMO.

+1

@ggetz is this a major change to this PR?  If we want this to go into 1.42, we'd like to merge it early next week so it has some time in master.

ggetz commented:
> @ggetz is this a major change to this PR? If we want this to go into 1.42, we'd like to merge it early next week so it has some time in master.

While getting it down to 0% when idle would take some more major changes (the viewer is still updating the clock, timeline with my approach), @kring's feedback has some good points I can get into this PR within a day or so that would get most of the way there. Thanks again @kring!

hpinkos commented:
Make sure you check that all the widgets still work.  I know you'll at least probably have to make updates to the `Cesium3DTilesInspector` and the `CesiumInspector`

ggetz commented:
Updated based on @kring's feedback.

> Make sure you check that all the widgets still work. I know you'll at least probably have to make updates to the Cesium3DTilesInspector and the CesiumInspector

I updated these two widgets. The other widgets do not modify the scene besides changes that automatically prompt rendering (camera updates, imagery and terrain changes).

There are some test failures after merging that I need to take a look at before merging.

ggetz commented:
@lilleyse or @hpinkos, I have the tests cleaned up. Can we take a final look at this?

mramato commented:
It looks like `scene.render` is now a request and may be a no-op.  If there is a reason I absolutely need to render, I have to call `scene.requestRender` immediately followed by `scene.render`, which seems a little verbose.  I can think of 3 solutions.

1. Add second "force" arg to scene.render to force the scene.
2. Add a `forceRender` function to scene which always renders
3. have `scene.render` always render and move the logic to the calling function (only the render loop should be calling scene.render anyway, right?)

2 or 3 would be my preference, does anyone else have any thoughts on this?

mramato commented:
> Besides simulation time changes and calls to requestRender, always render when:

With all of the additional changes, is this list still compelte?

> ImageryLayers are updated

What exactly does this mean?  When adding or removing an item from the imageryLayers collection?  changing the alpha/brigtness/gamma/etc..?

ggetz commented:
Thanks for taking a look @mramato!

> It looks like scene.render is now a request and may be a no-op...

I'm in favor of option 3

> With all of the additional changes, is this list still complete? What exactly does this mean? When adding or removing an item from the imageryLayers collection? changing the alpha/brigtness/gamma/etc..?

Updated and clarified the list in https://github.com/AnalyticalGraphicsInc/cesium/pull/6115#issue-288245822

mramato commented:
> Updated and clarified the list in #6115 (comment)

Thanks.  It might be a good idea to edit the original comment so users who look at this PR while trying to use the feature don't have out of date info.

Does assigning a new terrainProvider require a requestRender?  If we are handling a new imagery layer, changing terrainProvider should work as well, right?

ggetz commented:
> Does assigning a new terrainProvider require a requestRender? If we are handling a new imagery layer, changing terrainProvider should work as well, right?

Updated the comment to clarify, but the scene does not watch specifically for the terrain provider changing, just the new tiles being loaded and imagery layers changing. But because we're watching for those changes, changing providers will not require a request render.

mramato commented:
In case anyone is curious.  To update Cesium Ion's project editor (which is non-trivial), @hpinkos had to add about ~80 separate calls to `requestRender`, which seemed tedious to me at first but in truth they are almost all inside of property setters that are changing something about the scene.  So I think a well architected project should have no problem migrated to this new feature.

I'm a little worried over time this can lead to stray and seemingly random `requestRender` calls throughout the code, but I don't think that's going to end up happening (and even if there are some superfluous ones, it's a trivial function so it should not have a negative impact)

Thanks @ggetz! (And thanks @hpinkos for doing the migration work).

ggetz commented:
Updated. @hpinkos Is this good to go after the updates?

hpinkos commented on Source/Scene/Scene.js:
```diff
@@ -3038,13 +3066,7 @@ define([
-        }
-
-        if (shouldRender) {
+        if (shouldRender || cameraChanged) {
```
Don't you want to call your `shouldRender` function somewhere in here?  Otherwise this is always true

hpinkos commented:
I'm not crazy about this change.  Since the `requestRenderMode` option is on `Scene`, it should be `Scene`'s responsibility to figure out whether or not it should render.  This change moves that decision making to the Widget level, and will requires end users to do something extra if they're not using `CesiumWidget`.

hpinkos commented:
I would go with option 2 that @mramato proposed instead

ggetz commented:
I see your point @hpinkos. Updated.

hpinkos commented on Source/Scene/Scene.js:
```diff
@@ -3022,42 +3022,63 @@ define([
+            time = JulianDate.now();
+        }
+
+        updateAndRenderScene(this, time, true);
```
Instead of introducing another variable here, why not just `this._renderRequested = true`?

hpinkos commented on Source/Scene/Scene.js:
```diff
@@ -2848,44 +3007,88 @@ define([
+            time = JulianDate.now();
         }
+
+        updateAndRenderScene(this, time, true);
```
Instead of introducing another param here, why not just `this._requestRendered = true`?

hpinkos replied:
You probably don't need the private `updateAndRenderScene` function either.  You could just call `this.render()` from this function after setting `this._requestRendered = true`.  Unless I'm missing something.

hpinkos commented on Source/Scene/Scene.js:
```diff
@@ -2794,13 +2916,50 @@ define([
             Camera.clone(camera, scene._cameraClone);
-        } else if (scene._cameraStartFired && getTimestamp() - scene._cameraMovedTime > scene.cameraEventWaitTime) {
+
+            return true;
```
Relying on this function to determine whether or not the camera moved is going to be problematic. See #4753

ggetz replied:
Would you recommend using the `camera.changed` event instead? I can also update the `moveStart`/`moveEnd` events to fire according to that rather than using `cameraEqual`.

hpinkos replied:
Nope, `camera.changed` is broken for small camera changes as well.  We'll pretty much have to fix #4753 in order to determine if the camera changed.

ggetz replied:
@bagnell, do you know if there is a fix or a workaround we could use here?

bagnell replied:
We would need to fix #4753, but you can change the epsilon test to an equality test in `cameraEqual`.

mramato commented:
@ggetz can you clarify what the default behavior is for animation? Does it default to rendering every time the clock changes (that seems like the sensible default to me, I just wanted to be sure).

ggetz commented:
> can you clarify what the default behavior is for animation? Does it default to rendering every time the clock changes (that seems like the sensible default to me, I just wanted to be sure).

Updates occur each time the clock changes, then a render is triggered when the time since the last render is greater than `maximumRenderTimeChange`. The default is 0.5, but we can easily set this higher or lower. When I was testing, I often set this to `Infinity` so renders never happen with time changes. Setting to 0.0 would always update with time changes.

mramato commented:
> Setting to 0.0 would always update with time changes.

I definitely think 0 should be the default here. To be honest, I'm not sure why this property even needs to exist, since changing the animation speed with this value being 0 would have the same effect.  What use case are we trying to fulfill?

hpinkos commented:
Is `maximumRenderTimeChange` relying on system clock time or `Clock` time?

ggetz commented:
Edit:
> Setting to 0.0 would always ~~update~~ render with time changes.

> Is maximumRenderTimeChange relying on system clock time or Clock time?

Clock/simulation time.

> I definitely think 0 should be the default here. To be honest, I'm not sure why this property even needs to exist, since changing the animation speed with this value being 0 would have the same effect. What use case are we trying to fulfill?

Making this default to 0.0 would give you the same behavior as if you never enabled `requestRender` at all, because the clock advances every time the Viewer/Widget updates at the target frame rate. The point was that you can adjust the frequency at which renders would occur automatically. For example, if you want the stars to animate every so often, you do not need to render at the target frame rate unless you are interacting with the scene. If your scene content does not depend on the simulation time at all, setting this to `Infinity` would give you the best performance.

mramato commented:
@ggetz, I understand your use case and it does make sense, but if a user is animating it's most likely because they have content that is time dynamic and need it to update every frame.  It's very important that what my eyes see is not a lie compared to what the clock says (which is what happens now).  So for the common case I want an update every frame when I'm animating, and update on demand when I'm not.

Having a completely static scene but wanting the stars/sun/shadows to update once in a while seems like a far less common use case and should definitely be something that requires another level of opt-in.  Also, the rate at which those things change is so slow that they are bound to get updated anyway just through other auto-render triggers (like mouse movement).

mramato commented:
I could also be misunderstanding something, so if you want to chat offline, let me know.

ggetz commented:
After talking offline, @mramato , @hpinkos, and I came to the consensus that the best path forward is not NOT animate by default, and set the `maximumRenderTimeChange` default to `0`.

ggetz commented:
Opened https://github.com/AnalyticalGraphicsInc/cesium/pull/6154 for animation changes.

ggetz commented:
Also updated the default `maximumRenderTimeChange` and adjusted the camera equality test. It does not cause the events to be fired to often, and camera changes near and inside tilesets are now registered.

ggetz commented:
@hpinkos This is all updated, along with #6154

hpinkos commented:
Cool, I'll take a last look at this after #6154 is merged.  Does this fix #4753 now?

ggetz commented:
It should fix the `start` and `end` events, but it doesn't fix `camera.changed` which is mentioned at the bottom of the issue.

hpinkos commented:
@ggetz how that #6154 is merged, merge in master when you have a chance, then I'll review this

ggetz commented:
@hpinkos merged!

hpinkos commented:
@ggetz add a comment about fixing #4753 to CHANGES

hpinkos commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -0,0 +1,283 @@
+function pickingScenario() {
+    resetScene();
+
+    var entity = viewer.entities.add({
```
Call `requestRender` after adding this primitive

hpinkos replied:
Actually, that might not work since the entity is creating this primitive asynchronously.  How is an end user supposed to handle that case?

ggetz replied:
Since it is being created asynchronously, the return from the web worker will prompt a render automatically. Adding requestRender right after adding it to the scene would be best though, as it will ensure the primitive creation is kicked off.

hpinkos replied:
Ah okay, I thought it wasn't added until I moved the camera but I couldn't seem to reproduce it so maybe I just interpreted the timing wrong.

> Adding requestRender right after adding it to the scene would be best though, as it will ensure the primitive creation is kicked off.

I'm confused, will the web worker not start otherwise?

ggetz replied:
I can confirm that the entity does render without changing the camera. The call to primitives.update is still happens during a render since it's not trivial. So the geometry creation wont start until a render happens. This is similar to how we're loading in the globe tiles (see conversation above).

hpinkos commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -0,0 +1,283 @@
+var viewModel = {
+    requestRenderMode : true,
+    timeChangeEnabled : false,
+    maximumRenderTimeChange : 30.0,
```
This default is too high, I would think the common case would be  in the 0-1 range if time is moving forward.  Maybe make the slider between 1-5 or 1-10

ggetz replied:
I've adjusted the simulation time to be pretty high so the user can obviously see changes in the scene when a new render is requested. I can reset the clock and set this to a more common case if that would be more helpful.

hpinkos replied:
Um, I think the way the demo is set up right now just makes these options confusing.  You have to both start the animation and be on an example that has an animation to see what the difference is.  Can you add a new example and only show these panel options when that example is selected?  Use something like simple.czml that is always changing in an obvious way

hpinkos commented on Apps/Sandcastle/gallery/Scene Rendering Performance.html:
```diff
@@ -0,0 +1,283 @@
+});
+
+Cesium.knockout.getObservable(viewModel, 'timeChangeEnabled').subscribe(function(value) {
+    scene.maximumRenderTimeChange = value ? viewModel.maximumRenderTimeChange : undefined;
```
Why is the default for the `maximumRenderTimeChange` option passed to Viewer `Infinity` but this is being set to `undefined`?  Which is the right thing to do here?   From looking at the code, I guess it looks like they both do the same thing but I think it makes the example a little confusing.

hpinkos commented on Source/Widgets/CesiumWidget/CesiumWidget.js:
```diff
@@ -692,6 +696,11 @@ define([
         configureCameraFrustum(this);
+
+        var scene = this._scene;
+        if (scene.requestRenderMode) {
```
Is there any harm in calling `scene.requestRender()` if request render mode is off?  I would get rid of this if statement.

mramato commented:
Double clicking on an Entity does not cause a render, I would expect this to happen because it is part of the Viewer widget (and camera interaction)

mramato commented:
Same bug (probably same root cause) happens when you click on the camera icon on the InfoBox.

hpinkos commented:
> Double clicking on an Entity does not cause a render

Hmm, that's weird.  I would expect that to be caught with the camera change event since it's moving the camera and changing the transform

hpinkos commented:
Everything else looks fine to me!  Who else wants to review this before it's merged?  I'll plan on merging this first thing Monday morning unless someone wants more time to review (assuming all bugs are fixed).

ggetz commented:
> Hmm, that's weird. I would expect that to be caught with the camera change event since it's moving the camera and changing the transform

I would also expect the camera change to prompt a render, I'll take a look.

hpinkos commented:
I'm seeing a test failure
```
Scene/Scene scene morphing causes a new frame to be rendered in requestRenderMode
Expected JulianDate({ dayNumber: 2458144, secondsOfDay: 29660.149 }) not to equal JulianDate({ dayNumber: 2458144, secondsOfDay: 29660.149 }).
Error: Expected JulianDate({ dayNumber: 2458144, secondsOfDay: 29660.149 }) not to equal JulianDate({ dayNumber: 2458144, secondsOfDay: 29660.149 }).
    at stack (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1475:17)
    at buildExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1445:14)
    at Spec.expectationResultFactory (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:584:18)
    at Spec.addExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:332:34)
    at Expectation.addExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:528:21)
    at Expectation.toEqual (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1399:12)
    at http://localhost:8080/Specs/Scene/SceneSpec.js:1545:42
    at Object.<anonymous> (http://localhost:8080/Specs/spec-main.js:142:30)
    at attemptAsync (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1793:24)
    at QueueRunner.run (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1746:26)
```

ggetz commented:
@hpinkos Cleaned up the examples, and made the morphing specs more reliable.

Setting `trackedEntity` now prompts a render. This resolves the two bugs @mramato brought up, as they were setting tracked entity, but the camera didn't update until the next frame.

hpinkos commented:
I think you missed this one

> @ggetz add a comment about fixing #4753 to CHANGES

mramato commented:
I think we should enable `requestRenderMode` in CesiumViewer.js, since all it does is expose the Viewer widget and therefore works out of the box with the latest round of changes!

ggetz commented:
@hpinkos Thanks, fixed!

Add requestRenderMode to the CeisumViewer, and both build and unbuilt are working as is. 👍

hpinkos commented:
Taking a final look at this now, thanks for the reminder @ggetz!

hpinkos commented:
Great work on this @ggetz!

ggetz commented:
Thanks @hpinkos for all the reviews!

thw0rted commented:
Right now, when I toggle the `show` property for a DataSource in the `dataSources` collection, I have to explicitly request a render.  Would it make sense to automatically observe the `show` property when adding a source to the collection, and re-render any time it changes?  I guess you could actually watch at the `Entity.show` level, since ideally you'd re-render when any entity is shown or hidden, but I don't think there's a single easy event that fires any time an Entity that the Viewer "has" (in the `defaultDataSource` or somewhere in `dataSources`) is shown/hidden.

mramato commented:
@thw0rted While at a high level, I agree with you, the problem of trying to handle all of these cases for end users is that it's tedious to maintain and hard to scale.  I think one of the biggest scalability mistakes I made with the Entity API was trying to make everything automatically happen for the user.  The end result is that browsers gobble up memory handling all of the subscriptions and notifications.  So for request render mode, we made a conscious decision that the user would have to initiate the request.  We may revisit this in the future (and god knows I would love to completely refactor the Entity API) but I don't expect that to happen in the near future.

That being said, while writing this reply I had an idea.  You can actually just listen to the `EntityCollection.collectionChanged` event and call `requestRender` whenever it is fired.  I think that will give you exactly what you want and may actually be simple and robust enough to warrant a PR into master.  Do you have to give it a try and see how it works out for you?

Thanks.

thw0rted commented:
I had the same thought, having a sub to every entity's state would not scale very well, and I'm not sure how you even profile to figure out why it feels slow once it starts to bog down.  I'll take a look at `collectionChanged`, see if it's a better fit for my model.
